### PR TITLE
MAINT/DOC: refactor CircleCI config file

### PR DESCRIPTION
**Pull request by tupui**
In #22943 we saw that CircleCI was using the deployment keys in all jobs.

This PR update the configuration to restrict this on `main`. Doing that it also:

- Update to the latest configuration template
- Add a skip logic on the `circleci_artifacts_redirector_job`

cc @seberg @mattip

### CODE REVIEW

**Comment by tupui** [comment]
Somewhere some wrong syntax, looking at this now.

**Comment by tupui** [comment]
I will see to separate the build in multiple workflows running in parallel.

**Comment by seberg** [comment]
Parallelizing might be nice, but isn't a big deal. Especially for doc PRs it would mean quite a bit quicker preview, though!  I am happy to trust you on the setup and it looks like it does everything right.  However, right now it is failing in the last step?

**Comment by tupui** [comment]
Thanks for having a look @seberg, hopefully this will fix it.

Actually for the parallelism, why are there 2 jobs? Seems like it's the same job, one just fails with warnings while the other does not. So the artifact should be the same and we could just do the warnings job no? This is what we do for SciPy, one job. Or there is something I am missing?

File: .circleci/config.yml [hunk 1]
```diff
@@ -83,7 +87,7 @@ jobs:
             . venv/bin/activate
             cd doc
             # Don't use -q, show warning summary"
-            SPHINXOPTS="-j4 -n" make -e html || echo "ignoring errors for now, see gh-13114"
+            SPHINXOPTS="-j2 -n" make -e html || echo "ignoring errors for now, see gh-13114"
```
Comment (tupui) [inline .circleci/config.yml:90]: The machine on CircleCI only has 2 cores.

**Comment by seberg** [comment]
I don't remember.  I think the first job was set up to print all the warnings (and I guess fail due to that?).  That was mainly to help fix sphinx link errors IIRC.  I suppose that means if the first job doesn't fail, we indeed do not need the second one.

**Comment by tupui** [comment]
Ok I removed it. Although now that I think about this, then it will always be green and would not help. Ark so maybe the best is to have a parallel flow here. Or I just revert the last commit (CI is now green). LMK

**Comment by mattip** [comment]
> then it will always be green and would not help

Truthfully, for the few people who look at it, whether it is green or not doesn't really matter. I think it is fine to have 1 job

**Comment by mattip** [comment]
As long as we are here, is there a way to have a "docs only" build for changes in content to `doc/source`, that would only run a docs build and a doctest run?

**Comment by tupui** [comment]
Are you saying this? If we can cache the build and if we detect changes only in `doc/source`, then we don't build but use the cached version of the build? Should be doable, I need to look up how to detect the change only part.

**Comment by seberg** [comment]
Oh, I thought mattip meant a label or clearer tag then `[circle skip]`?  I guess if a change is really only in `doc/source` you could enable such a thing manually.

Maybe also OK to just do this if the "Documentation" tag (and none of the others) is used?  that would give the highest probability of actually being active.

In any case, I agree:  That should probably be a follow up.

**Comment by tupui** [comment]
Agreed for leaving this to some follow up. This needs a bit of thinking with the doc teams I suppose.

**Comment by mattip** [comment]
I meant that a change in doc/source/* does not require rebuilding most of the github, travis, and azure CI runs (I think the doctest ones are the only ones that still need to run). The ones here should still run. But yeah, it is too much of a corner case to worry about right now.

**Comment by tupui** [comment]
ok then this is ready on my side 😃

**Comment by seberg** [comment]
Lets give this a try then, thanks @tupui.

**Comment by tupui** [comment]
@seberg sorry the deploy step failed! Difficult to know what went wrong as I cannot SSH on CircleCI. You could re-run the job with SSH enabled and see the tree directory. It's for sure just a path issue.

File: .circleci/config.yml [hunk 2]
```diff
@@ -83,51 +87,50 @@ jobs:
...
+          paths:
+            - doc/build/html
+            - doc/neps/_build
+            - tools/ci/push_docs_to_repo.py
+
+  deploy:
+    <<: *defaults
+    steps:
+      - attach_workspace:
+          at: ~/
```
[the remark this answers is not in the document]
Comment (tupui) [reply .circleci/config.yml:117]: Probably that yes.

[the remark this answers is not in the document]
Comment (seberg) [reply .circleci/config.yml:117]: Lets just try it, thanks.

**Pull request merged**